server/src/main/java/org/elasticsearch/health/node/LocalHealthMonitor.java

```diff
@@ -91,63 +118,257 @@ private void registerListeners() {
+            currentMonitoring.cancel();
+        }
+    }
+
+    private void startMonitoringIfNecessary() {
         if (prerequisitesFulfilled && enabled) {
-            threadPool.scheduleUnlessShuttingDown(time, ThreadPool.Names.MANAGEMENT, this::monitorHealth);
+            if (isMonitorRunning() == false) {
+                monitoring = new Monitoring(
+                    monitorInterval,
+                    threadPool,
+                    lastReportedDiskHealthInfo,
+                    lastSeenHealthNode,
+                    diskCheck,
+                    clusterService,
+                    client
+                );
+                logger.debug("Local health monitoring started {}", monitoring);
+            } else {
+                logger.debug("Local health monitoring already started {}, skipping", monitoring);
+            }
         }
     }
 
-    // Helper method that starts the monitoring without a delay.
-    private void maybeScheduleNow() {
-        maybeScheduleNextRun(TimeValue.ZERO);
+    private boolean isMonitorRunning() {
+        Scheduler.Cancellable scheduled = this.monitoring;
+        return scheduled != null && scheduled.isCancelled() == false;
     }
 
     @Override
     public void clusterChanged(ClusterChangedEvent event) {
-        if (prerequisitesFulfilled == false) {
-            prerequisitesFulfilled = event.state().nodesIfRecovered().getMinNodeVersion().onOrAfter(Version.V_8_5_0)
-                && HealthMetadata.getFromClusterState(event.state()) != null;
-            maybeScheduleNow();
+        DiscoveryNode currentHealthNode = HealthNode.findHealthNode(event.state());
+        DiscoveryNode currentMasterNode = event.state().nodes().getMasterNode();
+        boolean healthNodeChanged = hasHealthNodeChanged(currentHealthNode, event);
+        boolean masterNodeChanged = hasMasterNodeChanged(currentMasterNode, event);
+        if (healthNodeChanged || masterNodeChanged) {
+            // On health node or on master node changes, the health node might be reset so the reported
+            // health info gets reset to null, to ensure it will be resent.
+            lastSeenHealthNode.set(currentHealthNode == null ? null : currentHealthNode.getId());
+            lastReportedDiskHealthInfo.set(null);
+            if (logger.isDebugEnabled()) {
+                String reason;
+                if (healthNodeChanged && masterNodeChanged) {
+                    reason = "the master node and the health node";
+                } else if (healthNodeChanged) {
+                    reason = "the health node";
+                } else {
+                    reason = "the master node";
+                }
+                logger.debug(
+                    "Resetting the health monitoring because {} changed, current health node is {}.",
+                    reason,
+                    currentHealthNode == null ? null : format("{{%s}{%s}}", currentHealthNode.getName(), currentHealthNode.getId())
```

Comment:
```suggestion
                    currentHealthNode == null ? null : format("[%s][%s]", currentHealthNode.getName(), currentHealthNode.getId())
```

Reply:
In most logging lines I have encountered we use this format `{{%s}{%s}}` is there a reason you want to switch to the `[%s][%s]` for this log line?

Reply:
Nvm, it's looks cleaner with the square brackets :)